brave/brave-core
# Ads pacing with pass-through-rate parameter in the catalog

Resolves https://github.com/brave/brave-browser/issues/11057

## Submitter Checklist:

- Verified that these changes build without errors on
- Verified that these changes pass automated tests (unit, browser, security tests) on

## Test Plan:

Unit tests cover probabilistic handling of pass-through rates.

For end-to-end testing:
- note the number of ads logged in brave://rewards-internals as `X eligible ads after pacing`
- adjust `Pass Through Rate` on the ads management system for one or more `house` type ads to `0`, once the ads catalog is refreshed by the client, make sure that the count logged has decreased by the number of ads edited
- adjust `Pass Through Rate` on the ads management system for one or more `house` type ads to `0.5`, once the ads catalog is refreshed by the client, wait for an ad to be shown 3-4 times, make sure that the count logged has changed on subsequent intervals without any other catalog changes

To test effects of pacing on delivery rate:
1. Note the number of ads delivered over the course of a day, hourly, for the release channel being tested
2. Adjust a group of campaigns with similar targeting criteria to have `Pass Through Rate` of 0.5
3. Collect the data over the course of another day
4. Compare the hourly delivery rates for each campaign to verify that it has decreased
5. time permitting, repeat steps 2-3 with `Pass Through Rate` of 0.75, verify the delivery rates are in between the previous days' values

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.
